chips/nrf52/src/usbd.rs

```diff
@@ -1563,16 +1605,26 @@ impl<'a> Usbd<'a> {
                 let size = regs.wlengthl.read(Byte::VALUE) + (regs.wlengthh.read(Byte::VALUE) << 8);
 
                 self.client.map(|client| {
+                    // Notify the client that the ctrl setup event has occurred.
+                    // Allow it to configure any data we need to send back.
                     match client.ctrl_setup(endpoint) {
                         hil::usb::CtrlSetupResult::OkSetAddress => {}
                         hil::usb::CtrlSetupResult::Ok => {
                             // Setup request is successful.
                             if size == 0 {
+                                // Directly handle a 0 length setup request.
                                 self.complete_ctrl_status();
                             } else {
                                 match regs.bmrequesttype.read_as_enum(RequestType::DIRECTION) {
                                     Some(RequestType::DIRECTION::Value::HostToDevice) => {
-                                        unimplemented!("CTRL write transaction");
+                                        // CTRL WRITE transfer with data to
+                                        // receive. We first need to setup DMA
+                                        // so that the hardware can write the
+                                        // data to us.
+                                        self.descriptors[endpoint]
+                                            .state
+                                            .set(EndpointState::Ctrl(CtrlState::WriteOut));
+                                        self.transmit_out_ep0();
```

Comment:
Here, the specification (https://infocenter.nordicsemi.com/pdf/nRF52840_OPS_v0.5.pdf page 618) invokes the `EP0RCVOUT` task **before** setting up the DMA and transmission. See my above comment, I think it would be much simpler if here we only invoke the `EP0RCVOUT` task, and just wait for the `EP0DATADONE` event.

Once `EP0DATADONE` has arrived, we can set up the DMA transfer like any other endpoint and launch it. See my comment in the `handle_ep0datadone` above.

This would be consistent with the other endpoints, and the sequence of events would be equivalent. Indeed, my understanding is that the specification proposes the sequence of events (1) `EP0RCVOUT` task (2) set up the DMA (3) `EP0DATADONE` event (4) `STARTEPOUT[0]` task.
But once the task (1) has been invoked, (2) and (3) are interchangeable because we have no control over when the USB controller with send the `EP0DATADONE` event. Even if we're fast at executing (2), the event (3) could still happen before we're done because this event happens in parallel in the USB controller. So we might as well wait for (3) to start executing (2), as we do for the other endpoints. What matters is that setting up the DMA in (2) happens before we invoke the task (4).

Reply:
That is not how I read "The software has to prepare EasyDMA by pointing to the buffer in Data RAM that shall contain the incoming data. If no other EasyDMA transfer is on-going with USBD, software can then send the EP0RCVOUT task, which will make the USBD to accept (ACK) the first OUT+DATA transaction from the host."

Reply:
Meh... I'm not convinced as to why the wording is different for the control endpoint vs. the bulk/interrupt endpoints.

But as I understand it:
- `EP0RCVOUT` task only makes the USBD accept (ACK) the first OUT+DATA transaction from the host. It's the equivalent of "writing any value to `SIZE.EPOUT[n]`" for bulk/interrupt endpoints.
- The DMA transfer won't start until the `STARTEPOUT[i]` task is invoked (for control and bulk/interrupt endpoints alike). So it's only necessary to setup EasyDMA before this point. But not necessarily before `EP0RCVOUT`.
- If one enables the `EP0DATADONE -> STARTEPOUT[0]` shortcut, then of course it's necessary to setup EasyDMA before sending `EP0RCVOUT`, because we have no control over when the `STARTEPOUT[0]` task is triggered. I think that was the spirit of the specs. But if we don't enable this shortcut, we can wait for the `EP0DATADONE` event before setting up the EasyDMA.
- If we wait for the `EP0DATADONE` event to setup EasyDMA, rather than directly after the `EP0SETUP` event, we block EasyDMA (`self.set_pending_dma()`) for a shorter amount of time, which is preferable.

To me, it would be much clearer to have a consistent implementation for all endpoints, rather than special casing the control endpoint beyond necessary.

Reply:
Doesn't it make sense then to leave the control flow that matches the datasheet to support a future developer who wants to add the shortcuts?

Reply:
As Tock we can make the choice not to use these shortcuts. IMO they add a bit of complexity (enabling/disabling them + a control flow that doesn't match the other endpoints) without clear benefits (the stack is a bit faster to reply, but it's only for control so is it necessary/noticeable?).

My suggestion would be to match what happens for the endpoints for the sake of clarity in Tock's code and to avoid code duplication. But of course one has to test whether that works.

TBH I haven't looked at the CDC protocol at all. Is there a test host program for this (in [tools/usb](https://github.com/tock/tock/tree/master/tools/usb) maybe)?